### DISCUSSION TITLE
How can I use HomeSpan to control my PowerPixels (= Powerful Neopixels) individually?

### DISCUSSION THREAD

**Discussion by FidelDworp** [HomeSpan/HomeSpan#558] [Ideas]
Since about 10 years, all lights in my home are controlled successfully by **Particle Photons** and **PowerPixels** (Photo below):

![PowerPixel_4 0-k](https://user-images.githubusercontent.com/46328609/231510414-94e9384d-cf99-4ff7-aae3-0a6cfa479473.jpeg)

These "daisy-chained" modules are essentially **NeoPixel chips** controlling each 3 MosFets for the led lamps and strips around my home. Via Homebridge I use Homekit with Siri. This works fine most of the time, but the Photon controllers have proven not so reliable. Therefore I’d like to replace this set-up by ESP32’s using HomeSpan. These PowerPiXel modules are incredibly useful as one can control a virtually unlimited number of lights or other devices from one GPIO pin, by simply plugging in more in a "daisy-chain".

I tried the Homespan example sketch “Pixel.ino”: I connected 8 Neopixels to pin 26 of an “ESP32 Dev Module” and it worked fine! Great stuff!
But this method is not useful for my purpose yet, as this sketch cannot address ‘particular pixels’, only strands of them and with the same colour settings.
So, I tried (with some help of ChatGPT ;-) to make a variation which creates a homekit object for a single addressable NeopiXel, but I did not yet find a way.

Here is the simplified set-up which goes part of the road, but not quite…

**Set-up**: I connect 3 Neopixels (# 0, 1, 2) to pin 25, Gnd and +5V.
I expect to get 3 objects in Homekit (Pixel1, Pixel2 and Pixel3), which should control the colour and brightness of neopixel # 0, 1, 2 respectively. When this exercise would work, I can see how to address any greater number of individual NeoPixels as needed.

**Sketch**: The code below works **partly** fine, but only the Homekit object “Pixel1” controls Neopixel (#0) ! 
The other objects in homekit (Pixel2 and Pixel3) don’t change the other leds (#1,2) because they are nowhere addressed in this sketch. I don't know if this is at all possible without modifying the Homespan library...

I would be very grateful if someone could help me realise this objective.

—
```cpp
// HomeSpan Addressable RGB LED Examples. Demonstrates use of:
//  * HomeSpan Pixel Class that provides for control of single-wire addressable RGB LEDs, such as the WS2812
// IMPORTANT: CHANGE PIN NUMBERS BELOW TO MATCH YOUR SPECIFIC ESP32/S2/C3 BOARD

// Setup for ESP32 Dev Module:
#if defined(CONFIG_IDF_TARGET_ESP32)
// Define 3 pixels on the same pin for 1-wire neopixels:
#define NEOPIXEL_RGB_PIN       25
#define DEVICE_SUFFIX          ""
#endif

#include "HomeSpan.h"
#include "extras/Pixel.h" // include the HomeSpan Pixel class

///////////////////////////////

struct NeoPixel_RGB : Service::LightBulb { // Addressable single-wire RGB LED Strand (NeoPixel)

  Characteristic::On power{0, true};
  Characteristic::Hue H{0, true};
  Characteristic::Saturation S{0, true};
  Characteristic::Brightness V{100, true};
  Pixel *pixel;
  uint8_t nPixels;

  NeoPixel_RGB(uint8_t pin, uint8_t nPixels) : Service::LightBulb() {
    V.setRange(5, 100, 1); // sets the range of the Brightness to be from a min of 5%, to a max of 100%, in steps of 1%
    pixel = new Pixel(pin); // creates Pixel LED on specified pin
    this->nPixels = nPixels; // save number of Pixels in this LED Strand
    update(); // manually call update() to set pixel with restored initial values
  }

  boolean update() override {

    int p = power.getNewVal();

    float h = H.getNewVal<float>(); // range = [0,360]
    float s = S.getNewVal<float>(); // range = [0,100]
    float v = V.getNewVal<float>(); // range = [0,100]

    Pixel::Color color;

    pixel->set(color.HSV(h * p, s * p, v * p), nPixels); // sets all nPixels to the same HSV color

    return (true);
  }
};

// Create the instances you want below!
void setup() {

  Serial.begin(115200);
  homeSpan.begin(Category::Lighting, "Pixel LEDS" DEVICE_SUFFIX);

  SPAN_ACCESSORY(); // create Bridge (note this sketch uses the SPAN_ACCESSORY() macro, introduced in v1.5.1 --- see the HomeSpan API Reference for details on this convenience macro)

  SPAN_ACCESSORY("Pixel1");
  new NeoPixel_RGB(NEOPIXEL_RGB_PIN, 1); // create first 1-LED NeoPixel with full color control

  SPAN_ACCESSORY("Pixel2");
  new NeoPixel_RGB(NEOPIXEL_RGB_PIN, 1); // create second 1-LED NeoPixel with full color control

  SPAN_ACCESSORY("Pixel3");
  new NeoPixel_RGB(NEOPIXEL_RGB_PIN, 1); // create third 1-LED NeoPixel with full color control
}

void loop() {
  homeSpan.poll();
}
```
—

Please note that I'm not a professional programmer: I only started programming for Arduino and Particles when I retired at 65... ;-)
 (I am teaching coding to my 8 grandchildren...)

Many thanks for helping!
Greetings from Belgium (EU)
FiDel

**Comment by HomeSpan** [comment]
Note that individual LEDs in a NeoPixel strand are not individually addressable.  They CAN all be different colors since you can SET the first LED to blue, the second to yellow, the third to light green etc.  But you can't directly address the third LED and set it to light green.  This is because the LEDs within a single strand are themselves all daisy-chained together.  To set the color of the first LED, you transmit a 24-bit color code.  To set both the first and second LED you transmit TWO 24-bit color codes.  The first one is saved by the first LED and the second one is passed from the first to the second LED for its use.  Thus, if you want to set the color of the third LED to light green, you need to transmit the color of the first, second, and third LED all at once.

HomeSpan does all of this automatically for you.  The Pixel command allows you to create an array of colors and with a single command it will transmit all the colors to a NeoPixel strand where the first color in the array will be used for the first LED, the second color in the array will be used for the second LED, etc.  There is also a variation of this method where you can set all the LEDs to the same color without needing to create a large array of the same color.  See the Pixel Page and examples for details on how these methods work.

Note if you connect three NeoPixel strands together, the result is a SINGLE strand with more LEDs.  There is no way to address just the LEDs of the second strand - they look simply like more LEDs attached to the first strand, so you would set the colors of the LED by simply creating a bigger array of colors and transmitting the whole pack together.  For example, if you connect two strands of 8 LEDs each and want the first set of LEDs to be red, and the second to be yellow, you'll need to create a single array of 8 reds and 8 yellows and then transmit the 16-color array to the strand.

An easier way would be separately connect the strands to DIFFERENT pins on the ESP32, and use separate instances of Pixel to control each strand.

Hope this helps provide some guidance for your project.

**Reply by FidelDworp** [reply]
I'm very happy with your quick response to my question, thanks!

It is surprising you say "_individual LEDs in a NeoPixel strand are not individually addressable_".
For your information, I include a simple sketch for an addressable led strip connected to a particle photon:
(It uses the <neopixel.h> library:  [https://github.com/adafruit/Adafruit_NeoPixel](url)  )

```cpp
#include <neopixel.h>

SYSTEM_MODE(AUTOMATIC);

// IMPORTANT: Set pixel PIN, COUNT and TYPE
#define PIXEL_PIN D2
#define PIXEL_COUNT 10
#define PIXEL_TYPE WS2811

Adafruit_NeoPixel pixels = Adafruit_NeoPixel(PIXEL_COUNT, PIXEL_PIN, PIXEL_TYPE);

int delayval = 500; // delay for half a second


void setup()
{
  pixels.begin(); // This initializes the NeoPixel library.
}


void loop()
{
  for(int i=0;i<PIXEL_COUNT;i++)   // For a set of NeoPixels the first NeoPixel is 0, second is 1, all the way up to the count of pixels minus one.
  {
    // pixels.Color takes RGB values, from 0,0,0 up to 255,255,255
    pixels.setPixelColor(i, pixels.Color(0,150,0)); // Moderately bright green color.
    pixels.show(); // This sends the updated pixel color to the hardware.
    delay(delayval); // Delay for a period of time (in milliseconds).
  }
    // Now only for the first 3 LEDS: Test the individual addressability!
    pixels.setPixelColor(1, pixels.Color(255,0,0));// RED
    pixels.show();
    delay(2000);
    
    pixels.setPixelColor(2, pixels.Color(0,255,0));// GREEN
    pixels.show();
    delay(2000);
    
    pixels.setPixelColor(3, pixels.Color(0,0,255)); // BLUE
    pixels.show();
    delay(2000);
    
    pixels.setPixelColor(1, pixels.Color(255,255,255)); // WHITE
    pixels.setPixelColor(2, pixels.Color(255,255,255)); // WHITE
    pixels.setPixelColor(3, pixels.Color(255,255,255)); // WHITE
    pixels.show();
    delay(2000);
    
    pixels.setPixelColor(1, pixels.Color(0,0,0)); // OFF
    pixels.setPixelColor(2, pixels.Color(0,0,0)); // OFF
    pixels.setPixelColor(3, pixels.Color(0,0,0)); // OFF
    pixels.show();
    delay(4000);
}
```

It is exactly this idea which brought me 10 years ago to develop my home lighting automation system with PowerPixels: 
I can set "one pixel" in a daisy chain to a particular RGB value, so that groups of lights (RGB or other) can be controlled in a very handy and simple way (via the MosFETs).

And when an individual Pixel is set to a different RGB value, none of the other Pixels colour value is affected. That is taken care of by this intelligent library...

Here is one node of my (10-node) home automation system:
![PowerPiXels](https://user-images.githubusercontent.com/46328609/233172747-b05c4bc2-6cd7-478e-ba9e-df35e660c3f9.jpg)

Of course, I could indeed use one ESP32 pin for each of the powerPixels, but that would mean that I need to rearrange all my powerPixels: Disconnect them from the daisy-chain (connected to only one pin) and connect each of them with 3 wires to the ESP32. That would make the whole set-up much more complex. But if Homespan will not be able to control Pixels individually, I will be forced to choose that route.

Here is my currently working Twin-strand of Neopixels with an ESP32:
![NeoPixel strips with ESP32](https://user-images.githubusercontent.com/46328609/233172668-2e63a324-b2e1-4bd6-881b-a4345c05ecc3.jpg)

I would gladly ship a number of "PowerPiXels" (and chocolates!?) to someone who can make this individual addressability work!
Homespan is a really fantastic system and I'm sad I didn't have it in 2013 to make my home automation project work with much less efforts and complexity... Great community also!

Greetz from Belgium

**Comment by HomeSpan** [comment]
For clarifications, the actual NeoPixel hardware - the physical strand - does not support addressing individual pixels.  If you have a strand of three pixels that are set to yellow and you want to change the first one to red, the software needs to send only a single color RED to change the first pixel.  But if instead you want to change only the **third** pixel to red, the software needs to send 3 colors - YELLOW, YELLOW, RED - to cause the strand to display yellow in the first two pixels and red in the third.

The Adafruit library you are using does exactly that.  It knows that is can't address individual pixels so it keeps an array of the colors for every pixel in memory.  When you call the method setPixelColor(), all that does is change the color in the internal array maintained by the Adafruit library.  Then, when you call the method `show()`, the library transmits every color in the array to the strand.  If you have a 50-pixel strand, it transmits 50 colors, even if you are only changing one.

In other words, the Adafruit library is simulating the ability for the end user to set a single pixel because it's doing all the work underneath.

The Pixel class in HomeSpan works the exact same way, but with slightly different commands.  You can create an array storing the colors of as many pixels as you want, set each pixel to whatever color needed, and then tell the Pixel class to upload that array in its entirety to the strand.  If you want to then change the color of one pixel, change just the color in the associated position in the array, and then send the array once again to the strand.  The Pixel page and example covers all the details on how to do this.

**Comment by FidelDworp** [comment]
Thanks again for responding!

Indeed, all the colours for each pixel must be sent every time.
I understood that already. 
But I cannot imagine how to do this with HomeSpan: I am not an educated programmer...

Can someone help me create a sketch I can try on a strand of 10 Neopixels? Creating also 10 objects in Apple Homekit App?

( PS: My offer for shipping PowerPiXels and chocolates from Belgium is still valid… 😉 )

Greetings,
FiDel

**Comment by HomeSpan** [comment]
Below is an example sketch that shows how to take a single 8-pixel strand and divide it into 3 subsets controlled individually.  Note if you have 3 strands of 10 pixels each and cascade them together to be attached to a single ESP32 pin, this looks just like a single 30-pixel strand.  You can create as many subsets as you want and a subset can cross the physical boundary from one pixel strand to another, as long as they are all cascaded together.

```C++
// THIS SKETCH SHOWS HOW A SINGLE NEOPIXEL STRAND (OR MULTIPLE STRANDS CASCADED TOGETHER TO ACT AS A SINGLE STRAND)
// CAN BE SPLIT INTO MULTIPLE SUBSETS OF PIXELS FOR INDIVIDUAL CONTROL

// IN THIS SKETCH WE USE AN 8-PIXEL STRAND WITH SEPARATE CONTROLS FOR PIXELS 1-3, PIXELS 4-5, AND PIXELS 6-8

#define NEOPIXEL_RGB_PIN       26           // PIN USED FOR NEOPIXEL STRAND (OR MULTIPLE STRANDS CASCADED TOGETHER)
#define NEOPIXEL_SIZE          8            // TOTAL NUMBER OF PIXELS IN NEOPIXEL STRAND
 
#include "HomeSpan.h"
#include "extras/Pixel.h"                   // include the HomeSpan Pixel class

// BELOW WE CREATE GLOBAL OBJECTS TO STORE ALL COLORS IN A SINGLE 8-PIXEL STRAND

Pixel pixelStrand(NEOPIXEL_RGB_PIN);                          // create a single, global instance of pixelStrand

Pixel::Color colorList[NEOPIXEL_SIZE]={Pixel::RGB(0,0,0)};    // create an array to hold all the pixel colors - initialize to all zeros

///////////////////////////////

struct NeoPixel_RGB : Service::LightBulb {      // Addressable single-wire RGB LED Strand (e.g. NeoPixel)
 
  Characteristic::On power{0,true};
  Characteristic::Hue H{0,true};
  Characteristic::Saturation S{0,true};
  Characteristic::Brightness V{100,true};

  uint8_t startingColor;
  uint8_t nColors;
  
  NeoPixel_RGB(uint8_t startingPixel, uint8_t nPixels) : Service::LightBulb(){

    V.setRange(5,100,1);                      // sets the range of the Brightness to be from a min of 5%, to a max of 100%, in steps of 1%
    this->startingColor=startingPixel-1;      // start of color array for this subset of pixels
    this->nColors=nPixels;                    // number of colors to change (which equals number of pixels in this subset)
    update();                                 // manually call update() to set pixel with restored initial values
  }

  boolean update() override {

    int p=power.getNewVal();
    
    float h=H.getNewVal<float>();       // range = [0,360]
    float s=S.getNewVal<float>();       // range = [0,100]
    float v=V.getNewVal<float>();       // range = [0,100]

    for(int i=0;i<nColors;i++)                                  // loop over a SUBSET of pixels and change their color
      colorList[i+startingColor]=Pixel::HSV(h*p, s*p, v*p);

    pixelStrand.set(colorList,NEOPIXEL_SIZE);            // refresh ALL pixels in the strand based on the updated colors
          
    return(true);  
  }
};

///////////////////////////////

void setup() {
  
  Serial.begin(115200);
 
  homeSpan.begin(Category::Lighting,"Pixel LEDS");

  SPAN_ACCESSORY();       // create Bridge (note this sketch uses the SPAN_ACCESSORY() macro, introduced in v1.5.1 --- see the HomeSpan API Reference for details on this convenience macro)
  
  SPAN_ACCESSORY("Light 1");    // Here we use the SPAN_ACCESSORY() macro and specify an optional name for the first set of LEDs
    new NeoPixel_RGB(1,3);      // create a LightBulb that controls pixels 1-3 of a Neo Pixel Strand (we start at pixel 1 with length of 3 pixels)

  SPAN_ACCESSORY("Light 2");    // Repeat for second set of LEDs
    new NeoPixel_RGB(4,2);      // create a LightBulb that controls pixels 4-5 of a Neo Pixel Strand (we start at pixel 4 with length of 2 pixels)

  SPAN_ACCESSORY("Light 3");    // Repeat for third set of LEDs
    new NeoPixel_RGB(6,3);      // create a LightBulb that controls pixels 6-8 of a Neo Pixel Strand (we start at pixel 6 with length of 3 pixels)
}

///////////////////////////////

void loop() {
  homeSpan.poll();
}

///////////////////////////////
```

**Reply by FidelDworp** [reply]
Wow, that sounds promising, thanks! That should cover all my needs, I believe.
I will only be able to test it when back home next week.

I'll keep you posted! 

Greetings,
FiDel

**Reply by FidelDworp** [reply]
I just had a chance to test your sketch and it works like a charm! I modified it to address pixel by pixel and also that works perfectly! Thanks again for your intervention!
PS: Can I contact you personally?

FiDel

**Reply by HomeSpan** [reply]
Really glad to hear it's working!  For questions/feedback unrelated to a specific HomeSpan function or issue, please feel free to email me at homespan@icloud.com.
